### ISSUE TITLE
`max-statements` `ignoreTopLevelFunctions` does not work with IIFE

### ISSUE THREAD

**Issue by JulioJu** [eslint/eslint#12950]
**Tell us about your environment**

$ yarn eslint --env-info
yarn run v1.22.0
$ /home/julioprayer/Blog/node_modules/.bin/eslint --env-info
Environment Info:

Node version: v13.9.0
npm version: v6.13.7
Local ESLint version: v6.8.0 (Currently used)
Global ESLint version: v6.8.0
Done in 6.08s.

**What parser (default, Babel-ESLint, etc.) are you using?**

`espree`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
env:
  browser: true

parser: espree

parserOptions:
  ecmaVersion: 6
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
'use strict';

/* eslint max-statements: ["error", 1, { "ignoreTopLevelFunctions": true }] */

// When I remove the following Immediately Invoked Function Expression, it works.
(function bar () {
  console.log("hello");
})();

function foo() {
  const foo1 = 1;
  const foo2 = 2;
}
```

```bash
$ yarn eslint ./scripts.js
```

**What did you expect to happen?**
No error reported

**What actually happened? Please include the actual, raw output from ESLint.**
```
$ npx eslint ./scripts.js

/home/julioprayer/Blog/scripts.js
  10:1  error  Function 'foo' has too many statements (2). Maximum allowed is 1  max-statements

✖ 1 problem (1 error, 0 warnings)
```
**Are you willing to submit a pull request to fix this bug?**
No, sorry

**Note**

***When I remove the Immediately Invoked Function Expression `bar`, it works.***

**Comment by mdjermanovic** [comment]
Hi @JulioJu, thanks for the issue!

This is working as intended, though this option should be probably clarified in the docs.

The option is designed to ignore functions that actually represent the top level. For example, in a script where all the code is wrapped in an IIFE, if the option is enabled the rule will ignore the number of statements in that IIFE. 

The option doesn't intend to ignore all functions defined _at_ the top level.

References: #4309, PR #4733

Technically, a function is considered to be the "top level" function in the given script if it isn't defined in another function, and all other functions are defined in that function. In the provided example, `bar` isn't defined in `foo`, so `foo` isn't considered to be the top level function.
